dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorAuthenticationProvider.java

```diff
@@ -128,6 +128,10 @@ public Authentication authenticate( Authentication auth )
                 throw new BadCredentialsException( "Invalid verification code" );
             }
         }
+        else if ( userCredentials.isTwoFA() && !(auth.getDetails() instanceof TwoFactorWebAuthenticationDetails) )
```

Comment:
Might read more clear when nesting the if-else in the first check `userCredentials.isTwoFA()`